File: backend/app/smart_api.py
"""Angel One SmartAPI v2.0 REST client.

Covers the four authenticated surfaces Delta-K needs: login-by-password (TOTP),
RMS margin, order placement and position/LTP reads.  Every call funnels through
:meth:`SmartApiClient._request`, which normalises Angel One's
``{status, message, errorcode, data}`` envelope into either a payload or a
:class:`SmartApiError`.
"""

from __future__ import annotations

import asyncio
import logging
from datetime import datetime
from typing import Any

import httpx

from .config import (
    LOGIN_URL,
    LTP_URL,
    PLACE_ORDER_URL,
    POSITION_URL,
    PROFILE_URL,
    REFRESH_URL,
    RMS_URL,
    settings,
)

log = logging.getLogger("deltak.smartapi")


class SmartApiError(RuntimeError):
    """Raised when SmartAPI answers with ``status: false`` or a transport error."""

    def __init__(self, message: str, code: str | None = None, raw: Any = None):
        super().__init__(message)
        self.code = code
        self.raw = raw


class SmartApiClient:
    """Stateful SmartAPI session holder.

    A single instance lives for the lifetime of the process; ``login`` may be
    called repeatedly (e.g. re-auth after token expiry) and is concurrency-safe.
    """

    def __init__(self) -> None:
        self.api_key: str = settings.api_key
        self.client_code: str = settings.client_code
        self.jwt_token: str | None = None
        self.refresh_token: str | None = None
        self.feed_token: str | None = None
        self.login_time: datetime | None = None
        self.state: str | None = None
        self.profile: dict | None = None
        self._client: httpx.AsyncClient | None = None
        self._lock = asyncio.Lock()

    # ------------------------------------------------------------------ #
    # plumbing
    # ------------------------------------------------------------------ #
    @property
    def authenticated(self) -> bool:
        return bool(self.jwt_token and self.feed_token)

    async def _http(self) -> httpx.AsyncClient:
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(
                timeout=httpx.Timeout(20.0, connect=10.0),
                headers={"User-Agent": "DeltaK-Terminal/1.0"},
            )
        return self._client

    async def aclose(self) -> None:
        if self._client and not self._client.is_closed:
            await self._client.aclose()

    def _headers(self, authorized: bool = True) -> dict[str, str]:
        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            "X-UserType": "USER",
            "X-SourceID": "WEB",
            "X-ClientLocalIP": settings.client_local_ip,
            "X-ClientPublicIP": settings.client_public_ip,
            "X-MACAddress": settings.mac_address,
            "X-PrivateKey": self.api_key,
        }
        if authorized:
            if not self.jwt_token:
                raise SmartApiError("Not authenticated — SmartAPI login required.")
            headers["Authorization"] = f"Bearer {self.jwt_token}"
        return headers

    async def _request(
        self,
        method: str,
        url: str,
        *,
        json_body: dict | None = None,
        authorized: bool = True,
        retries: int = 2,
    ) -> dict:
        client = await self._http()
        headers = self._headers(authorized=authorized)
        last_exc: Exception | None = None
        delay = 1.0

        for attempt in range(retries + 1):
            try:
                resp = await client.request(
                    method, url, headers=headers, json=json_body
                )
            except httpx.HTTPError as exc:  # transport failure — retry
                last_exc = exc
                if attempt < retries:
                    await asyncio.sleep(delay)
                    delay *= 2
                    continue
                raise SmartApiError(f"SmartAPI transport error: {exc}") from exc

            if resp.status_code >= 500 and attempt < retries:
                await asyncio.sleep(delay)
                delay *= 2
                continue

            try:
                payload = resp.json()
            except ValueError as exc:
                raise SmartApiError(
                    f"SmartAPI returned non-JSON (HTTP {resp.status_code})"
                ) from exc

            if resp.status_code == 401 or resp.status_code == 403:
                raise SmartApiError(
                    payload.get("message", "SmartAPI session expired"),
                    code=str(payload.get("errorcode")),
                    raw=payload,
                )

            if not payload.get("status", False):
                raise SmartApiError(
                    payload.get("message") or "SmartAPI request rejected",
                    code=str(payload.get("errorcode")),
                    raw=payload,
                )
            return payload

        raise SmartApiError(f"SmartAPI request failed: {last_exc}")

    # ------------------------------------------------------------------ #
    # auth
    # ------------------------------------------------------------------ #
    async def login(
        self,
        client_code: str,
        pin: str,
        totp: str,
        state: str | None = None,
    ) -> dict:
        """``POST /rest/auth/angelbroking/user/v1/loginByPassword``.

        ``totp`` is the six-digit code the user reads off their authenticator app
        at login time.  It is single-use and transient: it is forwarded straight
        to Angel One and never stored, logged or written to disk.  ``state`` is
        the optional passthrough string SmartAPI echoes back in the response.

        The API key comes from ``DK_API_KEY`` in the server environment, not from
        the caller.
        """
        totp = (totp or "").strip()
        if not totp:
            raise SmartApiError("A six-digit TOTP from your authenticator is required.")
        if not self.api_key:
            raise SmartApiError(
                "Server is missing DK_API_KEY — set the SmartAPI key in the "
                "deployment's secrets."
            )

        body: dict[str, str] = {
            "clientcode": client_code,
            "password": pin,
            "totp": totp,
        }
        if state:
            body["state"] = state

        async with self._lock:
            self.client_code = client_code
            payload = await self._request(
                "POST", LOGIN_URL, json_body=body, authorized=False, retries=1
            )
            data = payload.get("data") or {}
            self.jwt_token = data.get("jwtToken")
            self.refresh_token = data.get("refreshToken")
            self.feed_token = data.get("feedToken")
            self.state = data.get("state") or state
            if not self.jwt_token or not self.feed_token:
                raise SmartApiError("Login succeeded but tokens were missing.", raw=payload)
            self.login_time = datetime.now()
            log.info("SmartAPI session established for %s", client_code)

        try:
            self.profile = (await self.get_profile()).get("data")
        except SmartApiError:
            self.profile = None
        return data

    async def refresh_session(self) -> dict:
        """Re-mint the JWT with the stored refresh token."""
        if not self.refresh_token:
            raise SmartApiError("No refresh token on this session.")
        payload = await self._request(
            "POST",
            REFRESH_URL,
            json_body={"refreshToken": self.refresh_token},
            authorized=True,
            retries=1,
        )
        data = payload.get("data") or {}
        self.jwt_token = data.get("jwtToken") or self.jwt_token
        self.feed_token = data.get("feedToken") or self.feed_token
        self.refresh_token = data.get("refreshToken") or self.refresh_token
        return data

    def logout(self) -> None:
        self.jwt_token = None
        self.refresh_token = None
        self.feed_token = None
        self.profile = None
        self.login_time = None
        self.state = None

    # ------------------------------------------------------------------ #
    # authenticated reads
    # ------------------------------------------------------------------ #
    async def get_profile(self) -> dict:
        return await self._request("GET", PROFILE_URL)

    async def get_rms(self) -> dict:
        """``GET /rest/secure/angelbroking/user/v1/getRMS`` — margin snapshot."""
        payload = await self._request("GET", RMS_URL)
        return payload.get("data") or {}

    async def get_positions(self) -> list[dict]:
        payload = await self._request("GET", POSITION_URL)
        return payload.get("data") or []

    async def get_ltp(self, exchange: str, trading_symbol: str, token: str) -> dict:
        payload = await self._request(
            "POST",
            LTP_URL,
            json_body={
                "exchange": exchange,
                "tradingsymbol": trading_symbol,
                "symboltoken": token,
            },
        )
        return payload.get("data") or {}

    # ------------------------------------------------------------------ #
    # orders
    # ------------------------------------------------------------------ #
    async def place_order(
        self,
        *,
        trading_symbol: str,
        symbol_token: str,
        transaction_type: str,
        quantity: int,
        exchange: str = "NFO",
        order_type: str = "MARKET",
        product_type: str | None = None,
        duration: str = "DAY",
        price: float = 0.0,
        variety: str = "NORMAL",
        squareoff: float = 0.0,
        stoploss: float = 0.0,
    ) -> dict:
        """``POST /rest/secure/angelbroking/order/v1/placeOrder``."""
        body = {
            "variety": variety,
            "tradingsymbol": trading_symbol,
            "symboltoken": str(symbol_token),
            "transactiontype": transaction_type.upper(),
            "exchange": exchange,
            "ordertype": order_type.upper(),
            "producttype": (product_type or settings.product_type).upper(),
            "duration": duration,
            "price": f"{price:.2f}",
            "squareoff": f"{squareoff:.2f}",
            "stoploss": f"{stoploss:.2f}",
            "quantity": str(int(quantity)),
        }
        payload = await self._request("POST", PLACE_ORDER_URL, json_body=body, retries=0)
        return payload.get("data") or {}


#: Process-wide session used by the routers and the risk loop.
smart_api = SmartApiClient()

File: web/lib/server/smartapi.ts
import "server-only";

import { withRetry } from "@/lib/retry";
import { RateLimiter } from "./rateLimiter";
import {
  SESSION_COOKIE,
  SESSION_COOKIE_OPTIONS,
  SESSION_MAX_AGE,
  decodeSession,
  encodeSession,
  type ServerSession,
} from "./session";

// Re-exported so every existing `@/lib/server/smartapi` import keeps working
// now that the cookie envelope lives in `./session`, shared with Dhan.
export { SESSION_COOKIE, SESSION_COOKIE_OPTIONS, SESSION_MAX_AGE, decodeSession, encodeSession };
export type { ServerSession };

/**
 * Angel One SmartAPI v2.0 client for route handlers.
 *
 * Serverless functions are stateless, so there is no long-lived session object
 * here as there was in the Python engine. Instead:
 *
 *  - `DK_API_KEY` lives in the server environment and never leaves it for REST
 *    calls (it is separately exposed to the browser for the market-data
 *    WebSocket — see the login route for why, and what that does and does not
 *    grant).
 *  - The `jwtToken` — the credential that can place orders — is returned to the
 *    caller only inside an httpOnly cookie, so page JavaScript cannot read it.
 */

const API_ROOT = "https://apiconnect.angelone.in";
export const LOGIN_URL = `${API_ROOT}/rest/auth/angelbroking/user/v1/loginByPassword`;
export const REFRESH_URL = `${API_ROOT}/rest/auth/angelbroking/jwt/v1/generateTokens`;
export const PROFILE_URL = `${API_ROOT}/rest/secure/angelbroking/user/v1/getProfile`;
export const LOGOUT_URL = `${API_ROOT}/rest/secure/angelbroking/user/v1/logout`;
export const RMS_URL = `${API_ROOT}/rest/secure/angelbroking/user/v1/getRMS`;
export const PLACE_ORDER_URL = `${API_ROOT}/rest/secure/angelbroking/order/v1/placeOrder`;
export const LTP_URL = `${API_ROOT}/rest/secure/angelbroking/order/v1/getLtpData`;
export const CANDLE_URL = `${API_ROOT}/rest/secure/angelbroking/historical/v1/getCandleData`;
export const OI_DATA_URL = `${API_ROOT}/rest/secure/angelbroking/historical/v1/getOIData`;
export const PCR_URL = `${API_ROOT}/rest/secure/angelbroking/marketData/v1/putCallRatio`;
export const OI_BUILDUP_URL = `${API_ROOT}/rest/secure/angelbroking/marketData/v1/OIBuildup`;
export const MARGIN_BATCH_URL = `${API_ROOT}/rest/secure/angelbroking/margin/v1/batch`;

export const API_KEY = process.env.DK_API_KEY ?? "";
export const CLIENT_LOCAL_IP = process.env.DK_CLIENT_LOCAL_IP ?? "127.0.0.1";
export const CLIENT_PUBLIC_IP = process.env.DK_CLIENT_PUBLIC_IP ?? "127.0.0.1";
export const MAC_ADDRESS = process.env.DK_MAC_ADDRESS ?? "00:00:00:00:00:00";

/**
 * Angel One's own published throttling, per endpoint — request/second,
 * request/minute, request/hour where documented. `getOIData`,
 * `putCallRatio` and `OIBuildup` aren't in Angel One's published table; 1/sec
 * is a deliberately conservative default for them, not a documented number.
 *
 * Every call to a given endpoint shares one budget here regardless of which
 * browser tab or operator triggered it — see `rateLimiter.ts` for why that
 * has to live here rather than in the browser's own request queue.
 */
const limiters: Record<string, RateLimiter> = {
  [CANDLE_URL]: new RateLimiter([
    { ms: 1_000, max: 3 },
    { ms: 60_000, max: 150 },
    { ms: 3_600_000, max: 5_000 },
  ]),
  [OI_DATA_URL]: new RateLimiter([{ ms: 1_000, max: 1 }]),
  [PCR_URL]: new RateLimiter([{ ms: 1_000, max: 1 }]),
  [OI_BUILDUP_URL]: new RateLimiter([{ ms: 1_000, max: 1 }]),
  [MARGIN_BATCH_URL]: new RateLimiter([
    { ms: 1_000, max: 10 },
    { ms: 60_000, max: 500 },
    { ms: 3_600_000, max: 5_000 },
  ]),
};

export class SmartApiError extends Error {
  constructor(
    message: string,
    readonly status = 502,
    readonly code?: string,
  ) {
    super(message);
    this.name = "SmartApiError";
  }
}

export function smartApiHeaders(jwt?: string): Record<string, string> {
  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    Accept: "application/json",
    "X-UserType": "USER",
    "X-SourceID": "WEB",
    "X-ClientLocalIP": CLIENT_LOCAL_IP,
    "X-ClientPublicIP": CLIENT_PUBLIC_IP,
    "X-MACAddress": MAC_ADDRESS,
    "X-PrivateKey": API_KEY,
  };
  if (jwt) headers.Authorization = `Bearer ${jwt}`;
  return headers;
}

/** Angel One's throttle reply — `AB1004`, or a message that names the rate. */
function isRateLimit(message?: string, errorcode?: string): boolean {
  if (errorcode === "AB1004") return true;
  return /exceeding access rate|rate limit/i.test(message ?? "");
}

/** Normalise Angel One's `{status, message, errorcode, data}` envelope. */
export async function smartApiCall<T = Record<string, unknown>>(
  url: string,
  init: { method: "GET" | "POST"; body?: unknown; jwt?: string },
): Promise<T> {
  if (!API_KEY) {
    throw new SmartApiError(
      "Server is not configured with a SmartAPI key. Set DK_API_KEY in the deployment's environment.",
      503,
    );
  }

  // Waits its turn against every other call this deployment is making to the
  // same endpoint, from any request — not just this one. An endpoint with no
  // configured limiter (order placement, login, profile — none of which this
  // module's callers hit today) proceeds unthrottled.
  await limiters[url]?.acquire();

  let res: Response;
  try {
    // One retry, short gap: a bare `fetch()` failure here is Vercel's own
    // network to Angel One hiccupping, not a rate limit or a rejection (both
    // are classified from a real response below and never reach this catch)
    // — most of these clear in under a second and never need to surface as
    // the 502 the browser console otherwise shows for every one of them.
    //
    // The 8s AbortSignal is load-bearing, not decorative: a bare `fetch()`
    // has no timeout of its own, so one slow Angel One response used to hang
    // until Vercel's own hard function ceiling killed the whole request —
    // fatal for `/api/market/batch`, where up to 25 of these run per call
    // and one straggler took the entire ladder down with it instead of
    // failing just its own contract.
    res = await withRetry(
      () =>
        fetch(url, {
          method: init.method,
          headers: smartApiHeaders(init.jwt),
          body: init.body === undefined ? undefined : JSON.stringify(init.body),
          cache: "no-store",
          signal: AbortSignal.timeout(8_000),
        }),
      2,
      300,
    );
  } catch (err) {
    throw new SmartApiError(
      `SmartAPI transport error: ${err instanceof Error ? err.message : "unreachable"}`,
    );
  }

  let payload: {
    status?: boolean;
    message?: string;
    errorcode?: string;
    data?: unknown;
  };
  try {
    payload = await res.json();
  } catch {
    throw new SmartApiError(`SmartAPI returned non-JSON (HTTP ${res.status})`);
  }

  if (res.status === 401 || res.status === 403) {
    throw new SmartApiError(payload.message ?? "SmartAPI session expired", 401, payload.errorcode);
  }
  // Angel One meters the historical endpoints hard (a few requests a second per
  // key) and answers a breach with a 200 carrying an "exceeding access rate"
  // message. Surfacing that as 429 lets callers back off instead of treating a
  // throttle as a dead endpoint.
  if (res.status === 429 || isRateLimit(payload.message, payload.errorcode)) {
    throw new SmartApiError(
      payload.message ?? "SmartAPI rate limit exceeded",
      429,
      payload.errorcode,
    );
  }
  if (!payload.status) {
    throw new SmartApiError(
      payload.message ?? "SmartAPI request rejected",
      502,
      payload.errorcode,
    );
  }
  return (payload.data ?? {}) as T;
}

